[quant][pt2e] Add more precise representation for quantized add

jerryzh168 opened the pull request:
Summary:
The planned e2e for quantization in pytorch 2.0 export is the following:

float_model -> prepare_pt2e -> calibration -> convert_pt2e -> ...

inside convert_pt2e, we will first produce a q/dq representation of the quantized model, similar to the previous output of
convert_to_reference_fx in fx grah mode quantization:

```
torch.ops.quantized_decomposed.dequantize_per_tensor -> torch.ops.aten.add -> torch.ops.quantized_decomopsed.quantize_per_tensor
torch.ops.quantized_decomposed.dequantize_per_tensor   /
```

Then we'll rewrite the above to a more precise representation that express the intention in a more precise manner, since
here we actually want to do int8 addition, instead of simulating the int8 addition with fp32 operations, the representation for
quantized add is:

```
def quantized_add(x_i8, x_scale, x_zero_point, y_i8, y_scale, y_zero_point, out_scale, out_zero_point):
    x = (x_scale / out_scale) * x_i8
    y = (y_scale / out_scale) * y_i8
    out = x + y
    out -= (x_zero_point * x_scale - y_zero_point * y_scale) / out_scale
    out += out_zero_point
    return out
```

Test Plan:
```
buck2 test caffe2/test:quantization_pt2e -- --exact 'caffe2/test:quantization_pt2e - test_representation_add (quantization.pt2e.test_quantize_pt2e.TestQuantizePT2E)'
```

Reviewed By: kimishpatel

Differential Revision: D45628032



cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78

leslie-fang-intel commented on torch/ao/quantization/_pt2e/representation/rewrite.py:
```diff
@@ -0,0 +1,74 @@
+    y_i32 = y_i8.to(torch.int32)
+    # TODO: use out_dtype op
+    x_i32 = ((x_scale / out_scale) * (x_i32 - x_zero_point)).to(torch.int32)
+    y_i32 = ((y_scale / out_scale) * (y_i32 - y_zero_point)).to(torch.int32)
```
Hi @jerryzh168, thanks for providing this expression. Should we take care of the rounding issue here? Per my understanding, `to(torch.int32)` may directly trunk the decimal digits. Maybe we can do a `torch.round` firstly before converting to int32?

jerryzh168 replied:
this is temporary representation, it will be using out_dtype in the future after the PR is landed: https://github.com/pytorch/pytorch/pull/103333, but the implementation will be the same here since we'll have a general implementation of out_dtype that works for all use cases

kimishpatel replied:
THis is a good point. I agree with @leslie-fang-intel. Lets use torch.round? I think thats what we are doing for q/dq decomposition.

kimishpatel replied:
I think going from fp32 -> int32 should really use round.

jerryzh168 replied:
OK

jerryzh168 committed:
[quant][pt2e] Add more precise representation for quantized add (#104130)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/104130

The planned e2e for quantization in pytorch 2.0 export is the following:

float_model -> prepare_pt2e -> calibration -> convert_pt2e -> ...

inside convert_pt2e, we will first produce a q/dq representation of the quantized model, similar to the previous output of
convert_to_reference_fx in fx grah mode quantization:

```
torch.ops.quantized_decomposed.dequantize_per_tensor -> torch.ops.aten.add -> torch.ops.quantized_decomopsed.quantize_per_tensor
torch.ops.quantized_decomposed.dequantize_per_tensor   /
```

Then we'll rewrite the above to a more precise representation that express the intention in a more precise manner, since
here we actually want to do int8 addition, instead of simulating the int8 addition with fp32 operations, the representation for
quantized add is:

```
def quantized_add(x_i8, x_scale, x_zero_point, y_i8, y_scale, y_zero_point, out_scale, out_zero_point):
    x = (x_scale / out_scale) * x_i8
    y = (y_scale / out_scale) * y_i8
    out = x + y
    out -= (x_zero_point * x_scale - y_zero_point * y_scale) / out_scale
    out += out_zero_point
    return out
```

Test Plan:
```
buck2 test caffe2/test:quantization_pt2e -- --exact 'caffe2/test:quantization_pt2e - test_representation_add (quantization.pt2e.test_quantize_pt2e.TestQuantizePT2E)'
```

Reviewed By: kimishpatel

Differential Revision: D45628032

fbshipit-source-id: 9cbdd0cb398e6cfb97bf1a87526dc87ed6360621

jerryzh168 commented:
@pytorchbot merge
